### PR TITLE
Vtable

### PR DESCRIPTION
**Pull request by oli-obk**
This is a vtable impl that doesn't work for closures (that don't capture their environment)

It also doesn't work for functions that are casted to `Fn*` trait objects

### CODE REVIEW

File: src/interpreter/terminator.rs [hunk 1]
```diff
@@ -566,14 +575,14 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
-    is_provided: bool,
+pub (super) struct ImplMethod<'tcx> {
+    pub (super) method: Rc<ty::Method<'tcx>>,
+    pub (super) substs: &'tcx Substs<'tcx>,
+    pub (super) is_provided: bool,
 }
 
 /// Locates the applicable definition of a method, given its name.
-fn get_impl_method<'a, 'tcx>(
+pub (super) fn get_impl_method<'a, 'tcx>(
```
Comment (eddyb) [inline src/interpreter/terminator.rs]: What's with the space between `pub` and `(` everywhere? AFAIK it's non-canonical.

**Comment by eddyb** [comment]
LGTM as a first pass, thanks!

File: src/interpreter/vtable.rs [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+use rustc::ty::subst::{Substs, Subst};
+use rustc::ty;
+
+use super::EvalContext;
+use error::EvalResult;
+use memory::Pointer;
+use super::terminator::{get_impl_method, ImplMethod};
+
+impl<'a, 'tcx> EvalContext<'a, 'tcx> {
+    /// Creates a returns a dynamic vtable for the given type and vtable origin.
```
Comment (solson) [inline src/interpreter/vtable.rs]: `s/returns a //`

Comment (solson) [inline src/interpreter/vtable.rs]: Going to merge and fix this myself.

**Pull request merged**